python/src/utils.cpp

```diff
@@ -0,0 +1,59 @@
+
+#include "mlx/utils.h"
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <optional>
+
+namespace py = pybind11;
+using namespace py::literals;
+using namespace mlx::core;
+
+// Slightly different from the original, with python context on init we are not
+// in the context yet. Only create the inner context on enter then de   lete on
+// exit.
+class PyStreamContextManager {
+ public:
+  PyStreamContextManager(StreamOrDevice s) : _s(s), _inner(nullptr) {}
+
+  void enter() {
+    _inner = new StreamContextManager(_s);
+  }
+
+  void exit() {
+    if (_inner != nullptr) {
+      delete _inner;
+      _inner = nullptr;
+    }
+  }
+
+ private:
+  StreamOrDevice _s;
+  StreamContextManager* _inner;
+};
+
+void init_utils(py::module_& m) {
+  py::class_<PyStreamContextManager>(m, "StreamContextManager", R"pbdoc(
+        A context manager for setting the current device and stream.
+    
+        Example:
+        ```python
+        import mlx.core as mx
+    
+        # Create a context manager for the current device and stream.
+        with mx.StreamContextManager(mx.cpu):
+            # Run some code that uses the current device and stream.
+            pass
+        ```
```

Comment:
Could you move this to below the `Args` section?

Also you need to use an RST code block here:

```
```suggestion
        Example:
        
        .. code-block:: python
            
            import mlx.core as mx
    
            # Create a context manager for the current device and stream.
            with mx.StreamContextManager(mx.cpu):
                # Run some code that uses the current device and stream.
                pass
```